### PR TITLE
feat(spice): gate spice actors on runtime activation

### PR DESCRIPTION
**Pull request by viviveevee**
A spice build spawns the four spice actors unconditionally, but the chain they run
against may still be on a protocol version that predates spice activation.
This PR makes these actors process messages only when spice is activated:

- Every actor entry point asks whether spice is active before doing anything.
- The authority is the block's own header, not the epoch — `validate_header` now rejects
  any block whose spice-ness disagrees with its epoch, so `header.is_spice()` can be
  trusted.
- Block production reads the same epoch protocol version that `validate_header` checks
  against, so the new rule cannot reject an honestly produced block.

## Changes

- **`Chain::validate_header`** — the spice-ness check was one-way: a non-spice header in
  a spice epoch was rejected, a spice header in a pre-spice epoch was accepted. Now an
  equality check.

- **`chain/chain/src/spice/activation.rs`** (new) — the shared gate:
  - `spice_enabled_for_block` / `spice_enabled_at_head`.
  - `spice_enabled_at_head_on_startup` — no head yet is pre-spice, any other storage
    error is fatal.
  - `accept_spice_network_message` — for inbound peer traffic. Prefers the referenced
    block, falls back to the head when we do not hold its header, because spice
    legitimately receives data ahead of its block.

- **Actor gates** — one commit per actor, each returning early rather than acting on a
  pre-spice block:
  - *core writer* — `handle_processed_block`, `SpiceChunkEndorsementMessage`.
  - *chunk executor* — `handle_processed_block`, both crash-recovery steps in
    `start_actor`.
  - *chunk validator* — `ProcessedBlock`, contract accesses, contract code responses,
    state witnesses.
  - *data distributor* — `start_waiting_on_missing_data`, `ProcessedBlock`, incoming
    partial data, partial data requests, contract code requests. Data fetching is still
    scheduled either way.

  A node that sees activation while running needs no special handling: the gates open at
  the activation block and `ProcessedBlock` picks the work up from there.

- **Store and GC** — `AllNextBlockHashes` is no longer written for pre-spice blocks
  (every reader is spice-only). `clear_witnesses_data` gates on the build feature and
  then on the final head's spice-ness, rather than on `PROTOCOL_VERSION`.

- **`near_spice_pre_activation_messages_dropped_total{kind}`** — counts dropped messages
  by kind. Read against this node's head, so it grows while the head is pre-spice and is
  expected to stay flat once past activation.

- **Test utils** — `TestBlockBuilder` and the `new_dummy` constructors take an explicit
  protocol version instead of reading `PROTOCOL_VERSION`, so a spice build can produce
  pre-spice blocks.

## Testing

New suite in `test-loop-tests/src/tests/spice/pre_activation.rs`: builds a pre-spice
chain, counts every outbound network request, and fails on any unexpected inter-actor
message. Asserts the ten spice DB columns stay empty and the drop counters move, across
actor startup, kill/restart, `ProcessedBlock`, every network-routable message kind, and a
message naming a block we do not hold. `validate_header_spice.rs` covers the rejected
direction.

Un-ignores 17 pre-spice-pinned tests that were skipped under `protocol_feature_spice`
precisely because the actors used to trip over their chains:
`account_cost_increase_diff` (11), `deterministic_account_id` (2), `ml_dsa_verify`,
`p256_verify`, `reject_empty_method_name`, `sha3`.

## Known gap

Crossing the activation boundary is not addressed, only running on either side of it. The
two `fix_contract_loading_cost` tests stay ignored and their TODO now says why: the chunk
executor panics on the first spice block because the execution heads are never seeded at
activation. `accept_spice_network_message` carries the matching TODO on the network side.

### CODE REVIEW

File: chain/chain/src/garbage_collection.rs [hunk 1]
```diff
@@ -375,12 +376,26 @@ impl ChainStore {
...
+        if !cfg!(feature = "protocol_feature_spice") {
             return Ok(());
         }
 
         let final_head = self.final_head()?;
+        match spice_enabled_for_block(self, &final_head.last_block_hash) {
+            Ok(true) => {}
+            Ok(false) => return Ok(()),
+            Err(err) => {
+                tracing::debug!(
```
Comment (darioush) [inline chain/chain/src/garbage_collection.rs]: nit: should it be `warn!`?

File: chain/chain/src/spice/activation.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        }
+    }
+}
+
+/// Whether the block `block_hash` is a spice block.
+///
+/// Errors when the header is not on disk; callers that can be handed an
+/// arbitrary block hash by a peer should use [`accept_spice_network_message`],
+/// which folds that case into a drop decision.
+pub fn spice_enabled_for_block(
```
Comment (darioush) [inline chain/chain/src/spice/activation.rs:34]: nit: could consider using a name like `spice_enabled_for_header`? or `spice_enabled_for_block_header`

File: chain/chain/src/spice/activation.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            Self::ChunkEndorsement => "chunk_endorsement",
+            Self::PartialData => "partial_data",
+            Self::PartialDataRequest => "partial_data_request",
+            Self::ContractAccesses => "contract_accesses",
+            Self::ContractCodeRequest => "contract_code_request",
+            Self::ContractCodeResponse => "contract_code_response",
+            Self::StateWitness => "state_witness",
+        }
+    }
+}
```
Comment (darioush) [inline chain/chain/src/spice/activation.rs:27]: Both tables here can be derived. `strum` is already a non-optional dependency of `near-chain` (`chain/chain/Cargo.toml:27`) and of `test-loop-tests` (`test-loop-tests/Cargo.toml:36`).

`as_str` stays as a one-liner, so the four call sites and the `tracing` fields do not change.

```suggestion
use near_store::adapter::chain_store::ChainStoreAdapter;
use strum::{EnumIter, IntoStaticStr};

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, EnumIter, IntoStaticStr)]
#[strum(serialize_all = "snake_case")]
pub enum SpiceMessageKind {
    ChunkEndorsement,
    PartialData,
    PartialDataRequest,
    ContractAccesses,
    ContractCodeRequest,
    ContractCodeResponse,
    StateWitness,
}

impl SpiceMessageKind {
    pub fn as_str(self) -> &'static str {
        self.into()
    }
}
```

The one `ALL` caller becomes `SpiceMessageKind::iter()` with `use strum::IntoEnumIterator as _;`.

Precedent for the same shape: `ChunkRelevance` in `chain/client/src/stateless_validation/validate.rs:30` is a metric-label enum with `IntoStaticStr` + `serialize_all = "snake_case"`; `DBCol` in `core/store/src/columns.rs:20` pairs `EnumIter` with `IntoStaticStr`.

🤖 _-- Claude on behalf of Darioush_

File: chain/chain/src/spice/activation.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    block_hash: &CryptoHash,
+) -> bool {
+    let enabled = match spice_enabled_for_block(chain_store, block_hash) {
+        Ok(enabled) => enabled,
+        Err(_) => match spice_enabled_at_head(chain_store) {
+            Ok(enabled) => enabled,
+            // Neither the block nor the head is readable: we know nothing about
+            // this chain, so we cannot claim spice is active on it.
+            Err(err) => {
+                tracing::debug!(
```
Comment (darioush) [inline chain/chain/src/spice/activation.rs]: nit: consider using `error!` or `warn!`

File: test-loop-tests/src/tests/spice/pre_activation.rs [hunk 5]
```diff
@@ -0,0 +1,352 @@
+//! A spice-channel binary must run a chain whose protocol version predates spice
+//! activation. In a spice build the four spice actors are spawned unconditionally,
+//! so on a pre-spice chain they must stay inert at runtime.
```
Comment (darioush) [inline test-loop-tests/src/tests/spice/pre_activation.rs]: nit:
```suggestion
//! A spice-enabled binary must be able to run a chain whose protocol version predates spice
//! activation. In a spice build the spice actors are spawned unconditionally,
//! so before spice is enabled, they must stay inert at runtime.
```

File: test-loop-tests/src/tests/spice/pre_activation.rs [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+/// garbage collection run within the test; a larger window is needed when a node is
+/// killed and has to catch up afterwards.
+fn setup_pre_spice_chain(num_validators: usize, gc_num_epochs_to_keep: u64) -> TestLoopEnv {
+    let env = TestLoopBuilder::new()
+        .validators(num_validators, 0)
+        .epoch_length(EPOCH_LENGTH)
+        .protocol_version(pre_spice_protocol_version())
+        .protocol_upgrade_schedule(ProtocolUpgradeVotingSchedule::new_immediate(
+            pre_spice_protocol_version(),
+        ))
```
Comment (darioush) [inline test-loop-tests/src/tests/spice/pre_activation.rs:60]: nit: why is this needed?

File: test-loop-tests/src/tests/spice/pre_activation.rs [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+        assert_eq!(
+            self.spice.load(Ordering::Relaxed),
+            0,
+            "a spice actor sent a network request on a pre-spice chain",
+        );
+    }
+}
+
+fn dropped_count(kind: SpiceMessageKind) -> u64 {
+    SPICE_PRE_ACTIVATION_MESSAGES_DROPPED.with_label_values(&[kind.as_str()]).get()
```
Comment (darioush) [inline test-loop-tests/src/tests/spice/pre_activation.rs]: nit: ideally we don't need to look at the metrics for tests. right now each test gets a separate process with nexttest, so it is not a big deal. maybe some helper struct or field gated with `test_features` could be added?

File: test-loop-tests/src/tests/spice/pre_activation.rs [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+
+    fn install_on(&self, data: &mut TestLoopData, node: &NodeExecutionData) {
+        let counter = self.clone();
+        let peer_actor = data.get_mut(&node.peer_manager_sender.actor_handle());
+        peer_actor.register_override_handler(Box::new(move |request| -> HandlerResult {
+            counter.total.fetch_add(1, Ordering::Relaxed);
+            if is_spice_request(&request) {
+                counter.spice.fetch_add(1, Ordering::Relaxed);
+            }
+            HandlerResult::Unhandled(request)
```
Comment (darioush) [inline test-loop-tests/src/tests/spice/pre_activation.rs:166]: nit: can we add a comment that this passes the request through?

File: test-loop-tests/src/tests/spice/pre_activation.rs [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+        recv_permit: RecvMessagePermit::none(),
+    });
+    node_data.spice_data_distributor_sender.send(SpiceContractCodeRequestMessage(
+        SpiceContractCodeRequest::new(chunk_id.clone(), HashSet::new(), &signer),
+        RecvMessagePermit::none(),
+    ));
+
+    // The distributor forwards these two to the validator, which is where they are
+    // gated, so each is counted once at the end of a two-hop route the test does not
+    // have to replay by hand.
```
Comment (darioush) [inline test-loop-tests/src/tests/spice/pre_activation.rs:311]: nit: can we clarify here that SpiceChunkStateWitnessMessage has to be sent to the validator directly as it is an internal message?

File: chain/chain/src/metrics.rs [hunk 10]
```diff
@@ -91,6 +91,15 @@ pub static BLOCK_SPICE_UNCERTIFIED_CHUNKS: LazyLock<IntGauge> = LazyLock::new(||
     )
     .unwrap()
 });
+pub static SPICE_PRE_ACTIVATION_MESSAGES_DROPPED: LazyLock<IntCounterVec> = LazyLock::new(|| {
+    try_create_int_counter_vec(
+        "near_spice_pre_activation_messages_dropped_total",
+        "Number of spice messages dropped because the block they reference is not a spice \
+         block, by message kind",
```
Comment (darioush) [inline chain/chain/src/metrics.rs]: The help text covers one of the three paths that increment this counter. `accept_spice_network_message` also increments when the referenced block's header is not on disk and the head is pre-spice, and when neither is readable. In the first of those the referenced block may well be a spice block, which is the data loss the TODO at `activation.rs:85-87` describes.

```suggestion
        "Number of spice messages dropped because spice is not active for them, either \
         because the block they reference is not a spice block or because that block \
         could not be resolved, by message kind",
```

🤖 _-- Claude on behalf of Darioush_

File: chain/chain/src/spice/activation.rs [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+/// Whether spice is active at the current head.
+///
+/// For startup work, which has no particular block to key on, and as the
+/// fallback in [`accept_spice_network_message`].
+pub fn spice_enabled_at_head(chain_store: &ChainStoreAdapter) -> Result<bool, Error> {
+    let head = chain_store.head()?;
+    spice_enabled_for_block(chain_store, &head.last_block_hash)
+}
```
Comment (darioush) [inline chain/chain/src/spice/activation.rs:47]: nit: this is `ChainStoreAdapter::head_header()` (`core/store/src/adapter/chain_store.rs:93-99`), which does the same `head()?` then `get_block_header`.

```suggestion
pub fn spice_enabled_at_head(chain_store: &ChainStoreAdapter) -> Result<bool, Error> {
    Ok(chain_store.head_header()?.is_spice())
}
```

🤖 _-- Claude on behalf of Darioush_

File: chain/chain/src/spice/activation.rs [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+/// while a node past activation keeps buffering exactly as before.
+///
+/// TODO(spice): around the activation boundary this drops data about the first spice
+/// block that arrives before we hold that block's header, because the head is
+/// still pre-spice.
+pub fn accept_spice_network_message(
+    chain_store: &ChainStoreAdapter,
+    kind: SpiceMessageKind,
+    block_hash: &CryptoHash,
+) -> bool {
```
Comment (darioush) [inline chain/chain/src/spice/activation.rs]: nit: the name reads as a command, but this is a predicate that also increments a counter. `should_process_spice_network_message` states the question the four call sites are asking.

🤖 _-- Claude on behalf of Darioush_

File: test-loop-tests/src/tests/spice/pre_activation.rs [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+            );
+        }
+    }
+}
+
+/// Whether a network request carries spice traffic. Keyed on the variant name so that a
+/// spice request variant added later is counted without touching this test.
+fn is_spice_request(request: &NetworkRequests) -> bool {
+    request.as_ref().starts_with("Spice")
+}
```
Comment (darioush) [inline test-loop-tests/src/tests/spice/pre_activation.rs]: nit: all six spice variants of `NetworkRequests` start with `Spice` today (`chain/network/src/types.rs:316-326`), so this is correct now. But the doc claims a later variant is counted without touching this test, and that only holds if it is also named with the prefix. If it is not, `assert_no_spice_traffic` keeps passing while covering less.

An exhaustive `match` with no wildcard makes adding a variant a compile error here instead. (or an `is_spice_only()` impl.)

🤖 _-- Claude on behalf of Darioush_

File: core/primitives/src/test_utils.rs [hunk 14]
```diff
@@ -1328,11 +1340,13 @@ pub fn create_test_signer(account_name: &str) -> ValidatorSigner {
...
+pub fn pre_spice_protocol_version() -> ProtocolVersion {
+    ProtocolFeature::Spice.protocol_version() - 1
+}
+
 /// Build a minimal `ShardChunkHeaderV3` for use in tests that only need
 /// an object with a valid signature, `prev_block_hash`, and `shard_id`
-/// (everything else is zeroed). Mirrors the implicit defaults of
-/// `ShardChunkHeaderV3::new_dummy` but takes an explicit `protocol_version`
-/// so callers can pin the header inner variant.
+/// (everything else is zeroed).
```
Comment (darioush) [inline core/primitives/src/test_utils.rs]: nit: I think the `new_dummy` now takes the `protocol_version`

```suggestion
/// Build a minimal `ShardChunkHeaderV3` for use in tests that only need
/// an object with a valid signature, `prev_block_hash`, and `shard_id`
/// (everything else is zeroed). Takes an explicit `protocol_version` so
/// callers can pin the header inner variant.
```

🤖 _-- Claude on behalf of Darioush_

**Review by darioush — APPROVED** [review]
Nice! Just nits...

File: test-loop-tests/src/tests/spice/pre_activation.rs (hunk 6 above)
Comment (viviveevee) [reply test-loop-tests/src/tests/spice/pre_activation.rs:60]: Without it the chain runs at `PROTOCOL_VERSION`, which would be post-spice. Added a comment to explain

File: test-loop-tests/src/tests/spice/pre_activation.rs (hunk 7 above)
Comment (viviveevee) [reply test-loop-tests/src/tests/spice/pre_activation.rs]: Added a helper struct, doesn't do anything under non-`test_features`
